MultiLangProgGuide/MultiLangProgGuide.rst

```diff
@@ -625,14 +625,18 @@ that the transaction took too long to commit.
 ``YDB_ERR_UNKNOWN`` — A call to `ydb_message()`_ specified an
 invalid message code.
 
-``YDB_ERR_VARNAMEINVALID`` — A  variable name is too long. [#]_
+``YDB_ERR_VARNAMEINVALID`` — A  variable name is invalid. [#]_
```

Comment:
Note for exactly the reasons mentioned below (is the compiler's duty elsewhere), I had already added the INVVARNAME error. Also added the INVVARVALUE error. Here was how I defined those:

INVVARNAME	<Invalid local/global/ISV variable name/subscript supplied to API call>/error/fao=0
INVVARVALUE	<Invalid value specified to ydb_set_s() call>/error/fao=0

It is slightly different from yours so please pick/choose how you want it to be.

This validation takes place in the ydb_xxxx_s() routine itself because the actual runtime routines don't do validation. For example, the GVINVALID error seems to only raised by MUPIP when doing a LOAD command so GVINVALID is irrelevant to simpleAPI and does not intersect our project specifically.

The validation code I wrote is semi-simplistic so does not currently have several flavors of invalid name (missing name, bad character in name, bad subscript value, etc) and just returns INVVARNAME for all of them though if the subscript is missing (NULL pointer), the code lets op_putindx() give an UNDEF error instead though I may change that to raise a more specific error. You can go crazy with errors and I wouldn't mind some comments on how specific to get. For example, if a subscript is bad, how do you tell which one? Give a subscript index as part of the error? That will only help if they pull $ZSTATUS as that's the only place the error will have substitution values filled in.

Note these errors are not simpleAPI specific. They are being added to ydbmerrors so are defined like any other (new) error is.

Reply:
Since GVINVALID is the error raised by MUPIP for an invalid variable name, is there any reason for the Simple API to not raise the same GVINVALID error for an invalid variable name regardless of subscripts? For the name itself, any reason not to use a simple regular expression? The M pattern match ``?1(1AN,1"%")0.30AN`` detects a valid variable name, and we could use a pre-built state machine. To my thinking, an invalid variable name (or resource name for a lock) always results in GVINVALID, regardless of subscripts.

I'm OK with a separate SUBSINVALID error for invalid subscripts. It's OK to require a call to ydb_get_s() to get the get the value of $ZSTATUS to get specific error details - that's what it's there for.

Reply:
I think our error needs to be more generic because we are supporting local, global and ISV names with this interface. GVINVALID is not applicable to locals or ISVs. The error isn't really in the right format for us either since the error also includes the value (from a test reference file):

%GTM-E-GVINVALID, 	^e=1
			Invalid global name

So I think the INVVARNAME I have currently coded fits the bill unless you want to change it. I can change the errors for subscripts to INVSUBSCRIPT (I like the INV prefix as it denotes what follows is invalid making finding invalid messages easier - my own simple preference).

The code that does the local/global variable checking now is very simplistic and similar to what is used by the compiler now (and is done in macros). The ISV checking is the lookup of the name in a table to turn it into a "code" that the compiler normally does.

Reply:
OK to have a single, new error.

I have a slight preference for an object.message type of error name, e.g., VARNAMEINVALID rather than INVALIDVARNAME. That makes for easier indexing. But I'm willing to be persuaded otherwise. For what it's worth, I just polled Narayanan, and he prefers VARNAMEINVALID.

Reply:
As discussed, this will be INVALIDVARNAME because that's consistent with other messages.